crates/fuel-core/src/service/adapters.rs

```diff
@@ -85,6 +86,155 @@ impl StaticGasPrice {
+        let actual = subject
+            .worst_case_gas_price(target_height.into())
+            .await
+            .unwrap();
+
+        // then
+        let mut expected = gas_price;
+
+        for _ in 0..block_horizon {
+            let change_amount = expected
+                .saturating_mul(percentage as u64)
+                .saturating_div(100);
+            expected = expected.saturating_add(change_amount);
+        }
+
+        assert!(actual >= expected);
+    }
+
+    proptest! {
+        #[test]
+        fn worst_case_gas_price__correctly_calculates_value(
+            gas_price: u64,
+            starting_height: u32,
+            block_horizon in 0..10_000u32,
+            percentage: u16,
+        ) {
+            let rt = tokio::runtime::Runtime::new().unwrap();
+            rt.block_on(_worst_case__correctly_calculates_value(
+                gas_price,
+                starting_height,
+                block_horizon,
+                percentage,
+            ));
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn worst_case_gas_price__never_overflows(
+            gas_price: u64,
+            starting_height: u32,
+            block_horizon in 0..10_000u32,
+            percentage: u16
+        ) {
+            let rt = tokio::runtime::Runtime::new().unwrap();
+
+            // given
+            let subject = ArcGasPriceEstimate::new(starting_height, gas_price, percentage);
+
+            // when
+            let target_height = starting_height.saturating_add(block_horizon);
+
+            let _ = rt.block_on(subject.worst_case_gas_price(target_height.into()));
+
+            // then
+            // doesn't panic with an overflow
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub struct ArcGasPriceEstimate<Height, GasPrice> {
+    inner: Arc<parking_lot::RwLock<(Height, GasPrice)>>,
+    percentage: u16,
+}
+
+impl<Height, GasPrice> ArcGasPriceEstimate<Height, GasPrice> {
+    #[cfg(test)]
+    pub fn new(height: Height, price: GasPrice, percentage: u16) -> Self {
+        let pair = (height, price);
+        let inner = Arc::new(parking_lot::RwLock::new(pair));
+        Self { inner, percentage }
+    }
+
+    pub fn new_from_inner(
+        inner: Arc<parking_lot::RwLock<(Height, GasPrice)>>,
+        percentage: u16,
+    ) -> Self {
+        Self { inner, percentage }
+    }
+}
+
+impl<Height: Copy, GasPrice> ArcGasPriceEstimate<Height, GasPrice> {
+    pub fn get_height(&self) -> Height {
+        self.inner.read().0
+    }
+}
+impl<Height, GasPrice: Copy> ArcGasPriceEstimate<Height, GasPrice> {
+    pub fn get_gas_price(&self) -> GasPrice {
+        self.inner.read().1
+    }
+}
+
+#[async_trait::async_trait]
+impl GasPriceEstimate for ArcGasPriceEstimate<u32, u64> {
+    async fn worst_case_gas_price(&self, height: BlockHeight) -> Option<u64> {
+        let best_gas_price = self.get_gas_price();
+        let best_height = self.get_height();
```

Comment:
You can have a race condition here, because between `self.get_gas_price` and `self.get_height` the value can change. You can lock it, read both values, and unlock it again to avoid this issue.